src/math/fft/fft.c

```diff
@@ -0,0 +1,137 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2020 Intel Corporation. All rights reserved.
+//
+// Author: Amery Song <chao.song@intel.com>
+//	   Keyon Jie <yang.jie@linux.intel.com>
+
+#include <sof/audio/coefficients/fft/twiddle.h>
+#include <sof/audio/buffer.h>
+#include <sof/audio/format.h>
+#include <sof/common.h>
+#include <sof/lib/alloc.h>
+#include <sof/math/fft.h>
+
+struct fft_plan *fft_plan_new(struct icomplex32 *inb, struct icomplex32 *outb, uint32_t size)
+{
+	struct fft_plan *plan;
+	int lim = 1;
+	int len = 0;
+	int i;
+
+	if (!inb || !outb)
+		return NULL;
+
+	plan = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, sizeof(struct fft_plan));
+	if (!plan)
+		return NULL;
+
+	/* calculate the exponent of 2 */
+	while (lim < size) {
+		lim <<= 1;
+		len++;
+	}
+
+	plan->size = lim;
+	plan->len = len;
+
+	plan->bit_reverse_idx = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
+					plan->size * sizeof(uint32_t));
+	if (!plan->bit_reverse_idx) {
+		rfree(plan);
+		return NULL;
+	}
+
+	/* set up the bit reverse index */
+	for (i = 1; i < plan->size; ++i)
+		plan->bit_reverse_idx[i] = (plan->bit_reverse_idx[i >> 1] >> 1) |
+					   ((i & 1) << (len - 1));
+
+	plan->inb = inb;
+	plan->outb = outb;
+
+	return plan;
+}
+
+void fft_plan_free(struct fft_plan *plan)
+{
+	if (!plan)
+		return;
+
+	if (!plan->bit_reverse_idx)
+		rfree(plan->bit_reverse_idx);
```

Comment:
Would it be good to set the freed pointers to NULL? This and next one.

Reply:
it's also safe to pass NULL to rfree()

Reply:
> Would it be good to set the freed pointers to NULL? This and next one.

The pointer plan is from the caller so I better not to clear it, and since the whole plan struct will be freed so no need to do extra free of plan->bit_reverse_idx, but, but, I realize an logic error here, it should be:
```
--	if (!plan->bit_reverse_idx)
++	if (plan->bit_reverse_idx)
		rfree(plan->bit_reverse_idx);
```

Reply:
@singalsu updated.

Reply:
Again condition is unnecessary, again `rfree(NULL)` is a no-op and is safe to call.

Reply:
> Again condition is unnecessary, again `rfree(NULL)` is a no-op and is safe to call.

Oh correct, thanks for this, let me remove the condition.